### ISSUE TITLE
Unable to activate system prompts

### ISSUE THREAD

**Issue by tlindgren** [brianpetro/obsidian-smart-connections#568]
I was really excited to see the new system prompts feature but I can't seem to get it to work and I'm not sure if I'm doing something wrong. I have a system prompts folder saved and a note in the folder, but when I add the @ button in the chat, nothing happens. 

I'm using version 2.1.50.

Thanks

**Comment by brianpetro** [comment]
Hey @tlindgren 

Thanks for letting me know about this.

I think the issue is that nothing is indicating that nothing is indicating that the system prompt is being used in the Smart Chat window. I just made a note about this to get it fixed.

However, the system prompt should still be utilized by the chat. There are a couple ways that you can confirm this:

1) Click the button to the right of the name of the chat to open the chat history note. This note should include a "system" heading that contains the contents of your system prompt.

2) Inspect the request object being logged to the console.

Here's a screenshot you can use for reference:
![Screenshot 2024-04-26 at 11 26 00 PM](https://github.com/brianpetro/obsidian-smart-connections/assets/1886014/d8cf868d-ae0b-4752-93b2-f2daabe824c0)

If you confirm the system prompt is being included via either of the above steps but you feel like the chat isn't respecting the instructions, then it would be worth trying another chat model since some models pay more attention to system prompts that others.

🌴

**Comment by tlindgren** [comment]
Hi Brian, 

Thanks for getting back to me on this. Maybe I should back up to confirm how the feature is supposed to work. I'm assuming that after I designate a system prompt folder in settings, adding @ in the chat should pop up a list of my system prompts from that folder and then I can select the one I want to use. Do I have that correct? Or is there a different way I should be selecting a prompt. 

Tim

**Comment by brianpetro** [comment]
@tlindgren that is correct, it should look something like this 🌴
![Screenshot 2024-04-27 at 1 39 57 PM](https://github.com/brianpetro/obsidian-smart-connections/assets/1886014/735cffb6-7918-4d26-8512-a524f00d6ebf)

**Comment by brianpetro** [comment]
@tlindgren, thanks for the screenshot, that helps me better see what's going on. 

Are you copy & pasting the `@` in the video? 

What if you type it in without anything else in the input box, does it work then?

This feature is something that will be improved soon, but in the meantime, if you type the @ at the beginning without any other content it should work.

If it still doesn't work then, check the developer console because there should be an error there you can screenshot so we can figure it out.

🌴

**Comment by brianpetro** [comment]
@tlindgren those errors are unrelated and are OK.

Which operating system are you using?

Does typing `[[` open the file selection window?

Thanks for your help in solving this 
🌴

**Comment by tlindgren** [comment]
@brianpetro [[ and /folder/ both work fine. I'm Sonoma 14.4.1

**Comment by brianpetro** [comment]
@tlindgren interesting. I'm kind of stumped on this one then.

I'll let you know if I think of something. 

🌴

**Comment by PaulArthurM** [comment]
@brianpetro

Hi,

I had a similar issue on Windows 11. But for me, both '[[' and '@' weren't working.

 After some tries I found a workaround using 'keydown' instead of 'keyup'.

I'm not sure if this is the optimal solution, but it worked for me and may be a helpful starting point.

- I used the 'keydown' event listener instead of 'keyup' to detect key presses.
- I adjusted the conditions for checking the previous character when handling the '[[', '/', and '@' keys. I now check 'textarea.value[caret_pos - 1]' instead of 'textarea.value[caret_pos - 2]'.
- I added 'e.preventDefault()' to prevent the default behavior of the keys when the conditions for opening the suggestion modals are met.
- I modified the 'ScFileSelectModal', 'ScFolderSelectModal', and 'ScSystemPromptSelectModal' classes to include the corresponding characters ('[[', '/', '@') before the inserted text when an item is chosen from the suggestion modals.

```javascript
      add_chat_input_listeners() {
        super.add_chat_input_listeners();
        const chat_input = this.container.querySelector(".sc-chat-form");
        this.brackets_ct = 0;
        this.prevent_input = false;
        chat_input.addEventListener("keydown", this.key_down_handler.bind(this));
      }
      key_down_handler(e) {
        const textarea = this.container.querySelector(".sc-chat-form textarea");
        if (!["/", "@", "["].includes(e.key)) return;
        const caret_pos = textarea.selectionStart;
      
        if (e.key === "[") {
          if (textarea.value[caret_pos - 1] === "[") {
            e.preventDefault();
            this.open_file_suggestion_modal();
            return;
          }
        } else {
          this.brackets_ct = 0;
        }
      
        if (e.key === "/") {
          if (textarea.value.length === 0 || textarea.value[caret_pos - 1] === " ") {
            e.preventDefault();
            this.open_folder_suggestion_modal();
            return;
          }
        }
      
        if (e.key === "@") {
          if (textarea.value.length === 0 || textarea.value[caret_pos - 1] === " ") {
            e.preventDefault();
            this.open_system_prompt_modal();
            return;
          }
        }
      }
```

I also had to change this part as well to add the '@' or '[' character before the inserted text:

```javascript
    var ScFileSelectModal = class extends FuzzySuggestModal {
      constructor(app, env) {
        super(app);
        this.app = app;
        this.env = env;
        this.setPlaceholder("Type the name of a file...");
      }
      // get all markdown files
      getItems() {
        return this.app.vault.getMarkdownFiles().sort((a, b) => a.basename.localeCompare(b.basename));
      }
      getItemText(item) {
        return item.basename;
      }
      onChooseItem(file) {
        this.env.chat_ui.insert_selection("[" + file.basename + "]] ");
      }
    };
    var ScFolderSelectModal = class extends FuzzySuggestModal {
      constructor(app, env, folders) {
        super(app);
        this.app = app;
        this.env = env;
        this.folders = folders;
        this.setPlaceholder("Type the name of a folder...");
      }
      getItems() {
        return this.folders;
      }
      getItemText(item) {
        return item;
      }
      onChooseItem(folder) {
        this.env.chat_ui.insert_selection(folder + "/ ");
      }
    };
    var ScSystemPromptSelectModal = class extends FuzzySuggestModal {
      constructor(app, env) {
        super(app);
        this.app = app;
        this.env = env;
        this.setPlaceholder("Type the name of a system prompt...");
      }
      // getItems() { return this.env.system_prompts; }
      getItems() {
        return this.env.system_prompts;
      }
      getItemText(item) {
        return item.basename;
      }
      onChooseItem(prompt) {
        this.env.chat_ui.insert_selection('@"' + prompt.basename + '"');
      }
    };
  }
```

**Comment by linmaobang** [comment]
I succeeded, first to create a "smart prompts" folder, this folder name can be changed in the Settings, keep it same. can also create prompt command in this folder.Notice!!!!To use the @ command to bring up commands in the chat window, you need to hit the space bar before @! That's the key
![微信截图_20240516100215](https://github.com/brianpetro/obsidian-smart-connections/assets/5523824/35187628-3429-4ba0-8073-5f5fbe980eed)
![微信截图_20240516100239](https://github.com/brianpetro/obsidian-smart-connections/assets/5523824/6a9532d4-ad7c-4853-bbc2-31500e0ceecf)
![微信截图_20240516100302](https://github.com/brianpetro/obsidian-smart-connections/assets/5523824/7144452a-11f0-4bd7-ab8b-4c5422679b6d)
![微信截图_20240516100342](https://github.com/brianpetro/obsidian-smart-connections/assets/5523824/8179eced-58d9-4089-8a58-c263ded86cb1)

**Comment by Logonz** [comment]
I have this bug as well, no reaction when writing "@"/"<space>@" or "[[" or " [["
I am developer and looked into the console / behavior when trying to invoke but nothing seems to happen.
Is there any way to increase debug level to see if i can test it for you?

**Comment by Neeqstock** [comment]
Same problem.
Using Obsidian in Windows 11. Pressing @, / or [[ gives no reaction.

**Comment by tlindgren** [comment]
As @linmaobang mentioned hitting the space bar before @! worked for me as a workaround, but even that's a little finicky

**Comment by brianpetro** [comment]
@PaulArthurM Thanks for the detailed debugging insights! They played an important role in how I addressed this. The opening logic has been moved to the `keydown` handler. 

@tlindgren @Neeqstock @Logonz @linmaobang  While I wasn't able to reproduce this bug, based on @PaulArthurM's feedback, I believe it should now be fixed since `v2.1.70`

Thank you for everyone's patience and for bringing this to my attention 😊

🌴
